File: app/src/main/java/net/mercadosocial/moneda/util/update_app/UpdateAppManager.java
package net.mercadosocial.moneda.util.update_app;

import android.app.Activity;
import android.content.ActivityNotFoundException;
import android.content.Context;
import android.content.Intent;
import android.net.Uri;

import androidx.work.Constraints;
import androidx.work.ExistingPeriodicWorkPolicy;
import androidx.work.NetworkType;
import androidx.work.PeriodicWorkRequest;
import androidx.work.WorkManager;

import com.google.android.play.core.appupdate.AppUpdateInfo;
import com.google.android.play.core.appupdate.AppUpdateManager;
import com.google.android.play.core.appupdate.AppUpdateManagerFactory;
import com.google.android.play.core.appupdate.AppUpdateOptions;
import com.google.android.play.core.install.model.AppUpdateType;
import com.google.android.play.core.install.model.UpdateAvailability;
import com.google.android.play.core.tasks.Task;

import net.mercadosocial.moneda.BuildConfig;

import java.util.concurrent.TimeUnit;

public class UpdateAppManager {

    Context context;
    private AppUpdateManager appUpdateManager;
    private AppUpdateInfo appUpdateInfo;
    private UpdateAvailableListener updateAvailableListener;
    private UpdateFinishListener updateFinishListener;

    public static final String TEMPLATE_URL_GOOGLE_PLAY_APP_HTTP = "https://play.google.com/store/apps/details?id=%s";
    public static final String TEMPLATE_URL_GOOGLE_PLAY_APP_DIRECT = "market://details?id=%s";

    public UpdateAppManager(Context context) {
        this.context = context;

        appUpdateManager = AppUpdateManagerFactory.create(context);
    }

    public static void scheduleAppUpdateCheckWork(Context context) {

        Constraints constraints = new Constraints.Builder()
                .setRequiredNetworkType(NetworkType.CONNECTED)
                .build();

        PeriodicWorkRequest updateAppCheckWork = new PeriodicWorkRequest.Builder(UpdateAppCheckWorker.class,
                8, TimeUnit.HOURS,60, TimeUnit.MINUTES)
                .setConstraints(constraints)
                .build();

        WorkManager.getInstance(context).enqueueUniquePeriodicWork("appUpdateCheckWork",
                ExistingPeriodicWorkPolicy.KEEP, updateAppCheckWork);

    }

    public void setUpdateAvailableListener(UpdateAvailableListener updateAvailableListener) {
        this.updateAvailableListener = updateAvailableListener;
    }

    public void setFinishListener(UpdateFinishListener updateFinishListener) {
        this.updateFinishListener = updateFinishListener;
    }

    public void checkUpdateAvailable() {

        Task<AppUpdateInfo> appUpdateInfoTask = appUpdateManager.getAppUpdateInfo();

        appUpdateInfoTask
                .addOnCompleteListener(task -> {

                    if (task.isSuccessful()) {
                        appUpdateInfo = task.getResult();
                        if (appUpdateInfo.updateAvailability() == UpdateAvailability.UPDATE_AVAILABLE) {
                            if (updateAvailableListener != null) {
                                updateAvailableListener.onUpdateAvailable();
                            }
                        }

                    } else {
                        Exception e = task.getException();
//                        CountlyUtil.recordHandledException(e);
                        e.printStackTrace();

                    }

                    if (updateFinishListener != null) {
                        updateFinishListener.onCheckFinish();
                    }

                });
    }

    public void onUpdateVersionClick() {

        if (appUpdateInfo == null) {
            return;
        }

        if (appUpdateInfo.isUpdateTypeAllowed(AppUpdateType.IMMEDIATE)) {
            AppUpdateOptions options = AppUpdateOptions.newBuilder(AppUpdateType.IMMEDIATE)
                    .setAllowAssetPackDeletion(true).build();

            if (context instanceof Activity) {
                appUpdateManager.startUpdateFlow(appUpdateInfo, (Activity) context, options);
            }
        } else {
            openGooglePlay();
        }

    }

    private void openGooglePlay() {

        String packageName = BuildConfig.APPLICATION_ID.replace(".debug", "");
        String httpUrl = String.format(TEMPLATE_URL_GOOGLE_PLAY_APP_HTTP, packageName);
        String directUrl = String.format(TEMPLATE_URL_GOOGLE_PLAY_APP_DIRECT, packageName);

        Intent directPlayIntent = new Intent(Intent.ACTION_VIEW, Uri.parse(directUrl));
        try {
            context.startActivity(directPlayIntent);
        } catch (ActivityNotFoundException e) {
            context.startActivity(new Intent(Intent.ACTION_VIEW, Uri.parse(httpUrl)));
        }
    }

    public void onResume() {

        if (appUpdateInfo != null && appUpdateInfo.updateAvailability()
                == UpdateAvailability.DEVELOPER_TRIGGERED_UPDATE_IN_PROGRESS) {
            AppUpdateOptions options = AppUpdateOptions.newBuilder(AppUpdateType.IMMEDIATE)
                    .setAllowAssetPackDeletion(true).build();

            if (context instanceof Activity) {
                appUpdateManager.startUpdateFlow(appUpdateInfo, (Activity) context, options);
            }

        }
    }

    public interface UpdateAvailableListener {
        void onUpdateAvailable();
    }

    public interface UpdateFinishListener {
        void onCheckFinish();
    }
}

File: update-app-view/src/main/java/com/triskelapps/updateappview/UpdateAppManager.kt
package com.triskelapps.updateappview

import android.app.Activity
import android.content.ActivityNotFoundException
import android.content.Context
import android.content.Intent
import android.net.Uri
import android.util.Log
import androidx.work.Constraints
import androidx.work.ExistingPeriodicWorkPolicy
import androidx.work.NetworkType
import androidx.work.Operation
import androidx.work.PeriodicWorkRequest
import androidx.work.WorkManager
import com.google.android.gms.tasks.Task
import com.google.android.play.core.appupdate.AppUpdateInfo
import com.google.android.play.core.appupdate.AppUpdateManager
import com.google.android.play.core.appupdate.AppUpdateManagerFactory
import com.google.android.play.core.appupdate.AppUpdateOptions
import com.google.android.play.core.install.model.AppUpdateType
import com.google.android.play.core.install.model.UpdateAvailability

class UpdateAppManager(private val context: Context) {

    private val appUpdateManager: AppUpdateManager = AppUpdateManagerFactory.create(context)
    private var appUpdateInfo: AppUpdateInfo? = null
    private var onUpdateAvailable: () -> Unit = {}
    private var onUpdateFinish: () -> Unit = {}


    companion object {

        private val TAG: String = UpdateAppManager::class.java.simpleName

        const val TEMPLATE_URL_GOOGLE_PLAY_APP_HTTP: String =
            "https://play.google.com/store/apps/details?id=%s"
        const val TEMPLATE_URL_GOOGLE_PLAY_APP_DIRECT: String = "market://details?id=%s"

        lateinit var PACKAGE_NAME: String
        var VERSION_CODE: Int = 0

        var notificationIcon = 0
        var updateBarStyle: UpdateBarStyle? = null

        @JvmStatic
        @JvmOverloads
        fun init(
            context: Context,
            versionCode: Int,
            notificationIcon: Int,
            updateBarStyle: UpdateBarStyle? = null,
            checkWorkerConfiguration: CheckWorkerConfiguration? = CheckWorkerConfiguration()
        ) {
            PACKAGE_NAME = context.packageName
            VERSION_CODE = versionCode
            if (checkWorkerConfiguration != null) {
                scheduleAppUpdateCheckWork(context, checkWorkerConfiguration)
            }
            UpdateAppManager.notificationIcon = notificationIcon
            UpdateAppManager.updateBarStyle = updateBarStyle

            sendRemoteLog("UpdateAppManager init")
        }

        private fun scheduleAppUpdateCheckWork(
            context: Context?,
            checkWorkerConfiguration: CheckWorkerConfiguration
        ) {
            val constraints: Constraints = Constraints.Builder()
                .setRequiredNetworkType(NetworkType.CONNECTED)
                .build()

            val updateAppCheckWork: PeriodicWorkRequest = PeriodicWorkRequest.Builder(
                UpdateAppCheckWorker::class.java,
                checkWorkerConfiguration.repeatInterval, checkWorkerConfiguration.repeatIntervalTimeUnit,
                checkWorkerConfiguration.flexInterval, checkWorkerConfiguration.flexIntervalTimeUnit,
            )
                .setConstraints(constraints)
                .build()

            WorkManager.getInstance(context!!).enqueueUniquePeriodicWork(
                "appUpdateCheckWork",
                ExistingPeriodicWorkPolicy.UPDATE, updateAppCheckWork
            ).state.observeForever {state: Operation.State ->

                sendRemoteLog("Periodic work status: $state")
            }
        }
    }

    fun setUpdateAvailableListener(updateAvailableListener: () -> Unit = {}) {
        this.onUpdateAvailable = updateAvailableListener
    }

    fun setFinishListener(updateFinishListener: () -> Unit = {}) {
        this.onUpdateFinish = updateFinishListener
    }

    fun checkUpdateAvailable() {
        val appUpdateInfoTask: Task<AppUpdateInfo> = appUpdateManager.appUpdateInfo

        appUpdateInfoTask
            .addOnCompleteListener { task: Task<AppUpdateInfo?> ->
                if (task.isSuccessful) {
                    appUpdateInfo = task.result
                    if (appUpdateInfo!!.updateAvailability() == UpdateAvailability.UPDATE_AVAILABLE) {
                        onUpdateAvailable()
                    }
                } else {
                    Log.e(TAG, "checkUpdateAvailable: ", task.exception)
                }

                onUpdateFinish()

            }
    }

    fun onUpdateVersionClick() {
        if (appUpdateInfo != null) {
            if (appUpdateInfo!!.isUpdateTypeAllowed(AppUpdateType.IMMEDIATE)) {
                val options: AppUpdateOptions = AppUpdateOptions.newBuilder(AppUpdateType.IMMEDIATE)
                    .setAllowAssetPackDeletion(true).build()

                if (context is Activity) {
                    appUpdateManager.startUpdateFlow(appUpdateInfo!!, context as Activity, options)
                }
            } else {
                openGooglePlay()
            }
        }

    }

    private fun openGooglePlay() {
        val httpUrl = String.format(TEMPLATE_URL_GOOGLE_PLAY_APP_HTTP, PACKAGE_NAME)
        val directUrl = String.format(TEMPLATE_URL_GOOGLE_PLAY_APP_DIRECT, PACKAGE_NAME)

        val directPlayIntent = Intent(Intent.ACTION_VIEW, Uri.parse(directUrl))
        try {
            context.startActivity(directPlayIntent)
        } catch (e: ActivityNotFoundException) {
            context.startActivity(Intent(Intent.ACTION_VIEW, Uri.parse(httpUrl)))
        }
    }

    fun onResume() {
        if (appUpdateInfo != null && appUpdateInfo!!.updateAvailability()
            == UpdateAvailability.DEVELOPER_TRIGGERED_UPDATE_IN_PROGRESS
        ) {
            val options: AppUpdateOptions = AppUpdateOptions.newBuilder(AppUpdateType.IMMEDIATE)
                .setAllowAssetPackDeletion(true).build()

            if (context is Activity) {
                appUpdateManager.startUpdateFlow(appUpdateInfo!!, context, options)
            }
        }
    }


}
